src/avro_binary_decoder.erl

```diff
@@ -0,0 +1,331 @@
+  do_decode(iolist_to_binary(IoList), Type, LookupFun);
+do_decode(Bin, TypeName, LookupFun) when is_list(TypeName) ->
+  Type = LookupFun(TypeName),
+  do_decode(Bin, Type, LookupFun);
+do_decode(Bin, Type, LookupFun) ->
+  dec(Bin, Type, LookupFun).
+
+dec(Bin, T, _Lkup) when ?AVRO_IS_PRIMITIVE_TYPE(T) ->
+  prim(Bin, T#avro_primitive_type.name);
+dec(Bin, T, Lkup) when ?AVRO_IS_RECORD_TYPE(T) ->
+  FieldTypes = avro_record:get_all_field_types(T),
+  {FieldValuesReversed, Tail} =
+    lists:foldl(
+      fun({FieldName, FieldTypeOrName}, {Values, BinIn}) ->
+        {Value, BinOut} = do_decode(BinIn, FieldTypeOrName, Lkup),
+        {[{FieldName, Value} | Values], BinOut}
+      end, {[], Bin}, FieldTypes),
+  FieldValues = lists:reverse(FieldValuesReversed),
+  Name = avro:get_type_fullname(T),
+  {{Name, FieldValues}, Tail};
+dec(Bin, T, _Lkup) when ?AVRO_IS_ENUM_TYPE(T) ->
+  {Index, Tail} = int(Bin),
+  Symbol = avro_enum:get_symbol_from_index(T, Index),
+  {Symbol, Tail};
+dec(Bin, T, Lkup) when ?AVRO_IS_ARRAY_TYPE(T) ->
+  ItemsType = avro_array:get_items_type(T),
+  ItemDecodeFun = fun(BinIn) -> do_decode(BinIn, ItemsType, Lkup) end,
+  block(Bin, ItemDecodeFun);
+dec(Bin, T, Lkup) when ?AVRO_IS_MAP_TYPE(T) ->
+  ItemsType = avro_map:get_items_type(T),
+  ItemDecodeFun =
+    fun(BinIn) ->
+      {Key, Tail1} = prim(BinIn, "string"),
+      {Value, Tail} = do_decode(Tail1, ItemsType, Lkup),
+      {{Key, Value}, Tail}
+    end,
+  block(Bin, ItemDecodeFun);
+dec(Bin, T, Lkup) when ?AVRO_IS_UNION_TYPE(T) ->
+  {Index, Tail} = long(Bin),
+  {ok, ChildType} = avro_union:lookup_child_type(T, Index),
+  do_decode(Tail, ChildType, Lkup);
+dec(Bin, T, _Lkup) when ?AVRO_IS_FIXED_TYPE(T) ->
+  Size = avro_fixed:get_size(T),
+  <<Value:Size/binary, Tail/binary>> = Bin,
+  {Value, Tail}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+%% @private Decode primitive values.
+%% NOTE: keep all binary decoding exceptions to error:{badmatch, _}
+%%       to simplify higher level try catchs when detecting error
+%% @end
+prim(Bin, "null") ->
+  {null, Bin};
+prim(Bin, "boolean") ->
+  <<Bool:8, Rest/binary>> = Bin,
+  {Bool =:= 1, Rest};
+prim(Bin, "int") ->
+  int(Bin);
+prim(Bin, "long") ->
+  long(Bin);
+prim(Bin, "float") ->
+  <<Float:32/little-float, Rest/binary>> = Bin,
+  {Float, Rest};
+prim(Bin, "double") ->
+  <<Float:64/little-float, Rest/binary>> = Bin,
+  {Float, Rest};
+prim(Bin, "bytes") ->
+  bytes(Bin);
+prim(Bin, "string") ->
+  {Bytes, Tail} = bytes(Bin),
+  {unicode:characters_to_list(Bytes, utf8), Tail}.
+
+bytes(Bin) ->
+  {Size, Rest} = long(Bin),
+  <<Bytes:Size/binary, Tail/binary>> = Rest,
+  {Bytes, Tail}.
+
+block(Bin, ItemDecodeFun) -> block(Bin, ItemDecodeFun, _Acc = []).
+
+block(Bin, ItemDecodeFun, Acc) ->
+  {Count0, Rest} = long(Bin),
+  case Count0 =:= 0 of
```

Comment:
You parse trailing 0 as avro-binary-encoded integer but in encoder you use plain integer 0: https://github.com/klarna/erlavro/pull/4/files#diff-957b7df97c68b310ae983c3e82138ba4R107

Am I missing something?